### PR TITLE
Remove user-facing "reverse" property from Joint - part 1 of 2

### PR DESCRIPTION
**Pull request by tkuchida**
See #611.

### In this PR
- Removed `reverse` flag from Joint constructors
- Replaced Joint's `reverse` property with a protected member variable (`isReversed`)

### To be addressed in part 2
- Backwards compatibility via `updateFromXMLNode()`
- Remove `Joint::getIsReversed()`
- Update and add to testJoints.cpp

### CODE REVIEW

File: OpenSim/Simulation/SimbodyEngine/Joint.h [hunk 1]
```diff
@@ -529,21 +495,28 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Joint, ModelComponent);
...
+    //==========================================================================
     // DATA
-    //=========================================================================
+    //==========================================================================
+protected:
+    // Specifies the direction of the Joint in the multibody tree: parent->child
+    // (isReversed is false; default) or child->parent (isReversed is true). The
+    // Joint's transform and coordinates maintain a parent->child sense even if
+    // the joint is reversed.
+    bool isReversed = false;
```
Comment (chrisdembia) [inline OpenSim/Simulation/SimbodyEngine/Joint.h]: What about putting this description with the public accessor `isReversed()`? Might be good to also include a note about the history (`Before OpenSim 4.0, the user could set this. In 4.0 and later, this is determined automatically and cannot be set by the user`). I think the reason you put the description here is that this is supposed to be somewhat hidden. Perhaps the accessor's description can say "Advanced."

Comment (tkuchida) [inline OpenSim/Simulation/SimbodyEngine/Joint.h]: > do you think this deserves mention in the changelog?

We haven't been consistent with updating CHANGELOG; we will need to work on it prior to releasing 4.0 (see #384). But yes, I will add to CHANGELOG.

> What about putting this description with the public accessor `isReversed()`? Might be good to also include a note about the history ("Before OpenSim 4.0, the user could set this. In 4.0 and later, this is determined automatically and cannot be set by the user"). I think the reason you put the description here is that this is supposed to be somewhat hidden. Perhaps the accessor's description can say "Advanced."

I am going to remove `getIsReversed()` in part 2 when I improve the tests (`getIsReversed()` is used only by testJoints). The fact that joints can be reversed internally should be *completely* hidden from the user. I will add your note ("Before OpenSim 4.0, …") to CHANGELOG.

Comment (chrisdembia) [inline OpenSim/Simulation/SimbodyEngine/Joint.h]: > I am going to remove getIsReversed() in part 2 when I improve the tests (getIsReversed() is used only by testJoints). The fact that joints can be reversed internally should be completely hidden from the user. I will add your note ("Before OpenSim 4.0, …") to CHANGELOG.

Okay sounds good.

> We haven't been consistent with updating CHANGELOG

I know...I would like to create a [pull request template](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/) that reminds us to update the CHANGELOG, but I can't get over the fact that we would have to add another file or directory to the root of the project.

Comment (chrisdembia) [inline OpenSim/Simulation/SimbodyEngine/Joint.h]: Okay I am going to create a pull request template; I found a way to not add another entry to the root of the repo.

Comment (tkuchida) [inline OpenSim/Simulation/SimbodyEngine/Joint.h]: Addressed feedback:
- Added "Remove `Joint::getIsReversed()`" to "To be addressed in part 2" list
- Added note to CHANGELOG

Comment (aseth1) [inline OpenSim/Simulation/SimbodyEngine/Joint.h]: > I am going to remove getIsReversed() in part 2 when I improve the tests (getIsReversed() is used only by testJoints). The fact that joints can be reversed internally should be completely hidden from the user.

I completely agree!

**Comment by aseth1** [comment]
Thanks @tkuchida for cleaning this up. Merging ...

**Comment by tkuchida** [comment]
Thanks for revierging.

**Pull request merged**